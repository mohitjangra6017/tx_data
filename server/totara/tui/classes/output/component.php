<?php
/*
 * This file is part of Totara Learn
 *
 * Copyright (C) 2019 onwards Totara Learning Solutions LTD
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * @author Simon Chester <simon.chester@totaralearning.com>
 * @package totara_core
 */

namespace totara_tui\output;

/**
 * Renderable TUI component
 */
final class component implements \renderable {
    /**
     * @var string Component name
     */
    private $name;

    /**
     * @var array|null Component props
     */
    private $props = [];

    /**
     * @var bool
     */
    private static $registered = [];

    /**
     * Create a new instance of component
     *
     * @param string $name Module ID of the component to render.
     *     e.g. 'totara_core/components/Example`.
     * @param array|null $props Props to pass to the component.
     *     Props are not processed or escaped in any way, they are passed as-is
     *     to the component. Must be encodable as JSON.
     */
    public function __construct(string $name, array $props = null) {
        $this->name = $name;
        $this->props = $props;
    }

    /**
     * Get component name
     *
     * @return string
     */
    public function get_name(): string {
        return $this->name;
    }

    /**
     * Get component props
     *
     * @return array Mapping of prop keys to values
     */
    public function get_props(): ?array {
        return $this->props;
    }

    /**
     * @return bool
     */
    public function has_props(): bool {
        return ($this->props !== null);
    }

    /**
     * @throws \coding_exception if there are problems json encoding data for the component.
     * @return string
     */
    public function get_props_encoded(): string {
        if (!$this->has_props()) {
            // If you get here and want to remove this then you must work out and confirm what empty props look like for Vue.
            // It wasn't required during development and it's a bad idea to support something you don't need!
            throw new \coding_exception('Encoded props requested, but there are no props.');
        }
        $data = json_encode($this->props);
        if ($data === false && json_last_error() !== JSON_ERROR_NONE) {
            throw new \coding_exception('Invalid component data encountered while attempting to encode a component of type ' . static::class, json_last_error_msg());
        }
        return $data;
    }

    public function register(\moodle_page $page): \moodle_page {
        return self::register_component($this->get_name(), $page);
    }

    public static function register_component(string $name, \moodle_page $page): \moodle_page {
        $pos = strpos($name, '/');
        if ($pos !== false) {
            $name = substr($name, 0, $pos);
        }
        if (!isset(self::$registered[$name])) {
            if ($page->state >= \moodle_page::STATE_IN_BODY) {
                throw new \coding_exception('Unable to register component as the header has already been printed.');
            }
            self::get_framework($page)->require_vue($name);
            self::$registered[$name] = true;
        }
        return $page;
    }

    /**
     * @param \moodle_page $page
     * @throws \coding_exception
     * @return framework
     */
    private static function get_framework(\moodle_page $page): framework {
        /** @var framework $framework */
        $framework = $page->requires->framework(framework::class);
        return $framework;
    }
}
