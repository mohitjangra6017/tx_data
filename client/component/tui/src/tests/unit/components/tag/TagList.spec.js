/**
 * This file is part of Totara Enterprise Extensions.
 *
 * Copyright (C) 2020 onwards Totara Learning Solutions LTD
 *
 * Totara Enterprise Extensions is provided only to Totara
 * Learning Solutions LTD's customers and partners, pursuant to
 * the terms and conditions of a separate agreement with Totara
 * Learning Solutions LTD or its affiliate.
 *
 * If you do not have an agreement with Totara Learning Solutions
 * LTD, you may not access, use, modify, or distribute this software.
 * Please contact [licensing@totaralearning.com] for more information.
 *
 * @author Tatsuhiro Kirihara <tatsuhiro.kirihara@totaralearning.com>
 * @module tui
 */

import { mount, shallowMount } from '@vue/test-utils';
import component from 'tui/components/tag/TagList.vue';
let wrapper;

const stubs = {
  Close: true,
  InputText: true,
  Expand: true,
  OverflowDetector: {
    render() {
      return this.$scopedSlots.default({ measuring: true });
    },
  },
};

const data = function() {
  return {
    visible: 0,
  };
};

const propsData = {
  tags: [{ id: 1, text: 'Tui' }],
};

describe('presentation/tag/TagList.vue', () => {
  beforeAll(() => {
    wrapper = mount(component, {
      stubs,
      data,
      propsData,
    });
  });

  it('Checks snapshot', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('closeOnClick is false by default', () => {
    expect(wrapper.vm.closeOnClick).toBe(false);
  });

  it('closeOnClick is passed through dropdown', () => {
    wrapper = shallowMount(component, {
      stubs,
      data,
      propsData: Object.assign({}, propsData, { closeOnClick: true }),
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('uses the labelName provided for dropdown button, input field and remove tag button', () => {
    wrapper = mount(component, {
      propsData: Object.assign({}, propsData, {
        labelName: 'Learning category',
      }),
    });
    expect(wrapper.element).toMatchSnapshot();
  });
});
