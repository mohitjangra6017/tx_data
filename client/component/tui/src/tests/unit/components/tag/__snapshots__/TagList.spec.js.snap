// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`presentation/tag/TagList.vue Checks snapshot 1`] = `
<div
  class="tui-dropdown tui-dropdown--flexibleWidth"
>
  <div
    class="tui-dropdown__trigger"
    id="id-trigger"
  >
    <div
      class="tui-tagList"
    >
      <div
        class="tui-tagList__tags"
      >
        <div
          aria-atomic="false"
          aria-label="[[tags_selected, totara_core]]"
          aria-live="polite"
          aria-relevant="additions"
          class="tui-tagList__tagItems"
          role="status"
        >
          <div
            class="tui-tag"
          >
            <span
              class="tui-tag__content tui-tag__content--withButton"
            >
              Tui
            </span>
             
            <button
              aria-label="[[tag_remove, totara_core, \\"Tui\\"]]"
              class="tui-iconBtn tui-iconBtn--prim tui-iconBtn--small tui-iconBtn--transparent"
              title="[[tag_remove, totara_core, \\"Tui\\"]]"
              type="button"
            >
              <span
                class="tui-iconBtn__wrap"
              >
                <span
                  class="tui-iconBtn__label"
                >
                  <span
                    aria-hidden="true"
                    class="tui-iconBtn__icon"
                  >
                    <close-stub
                      size="100"
                    />
                  </span>
                   
                  <!---->
                </span>
                 
                <!---->
                 
                <!---->
              </span>
            </button>
          </div>
           
          <div
            class="tui-tagList__input"
          >
            <inputtext-stub
              arialabel="[[filter_x_taglist, totara_core, \\"[[tag_list, totara_core]]\\"]]"
              styleclass="[object Object]"
              value=""
            />
          </div>
        </div>
         
        <span
          class="tui-tagList__suffix"
        >
          [[n_more, totara_core, 1]]
        </span>
      </div>
       
      <button
        aria-expanded="false"
        aria-haspopup="menu"
        aria-label="[[tag_list, totara_core]]"
        class="tui-iconBtn tui-tagList__expandArrow tui-iconBtn--transparent"
        title="[[tag_list, totara_core]]"
        type="button"
      >
        <span
          class="tui-iconBtn__wrap"
        >
          <span
            class="tui-iconBtn__label"
          >
            <span
              aria-hidden="true"
              class="tui-iconBtn__icon"
            >
              <expand-stub
                customclass="tui-tagList__caret"
                size="200"
              />
            </span>
             
            <!---->
          </span>
           
          <!---->
           
          <!---->
        </span>
      </button>
    </div>
  </div>
   
  <!---->
</div>
`;

exports[`presentation/tag/TagList.vue closeOnClick is passed through dropdown 1`] = `
<dropdown-stub
  animation="default"
  canclose="true"
  closeonclick="true"
  contextmode="contained"
  matchwidth="true"
  position="bottom-left"
  role="menu"
>
   
</dropdown-stub>
`;

exports[`presentation/tag/TagList.vue uses the labelName provided for dropdown button, input field and remove tag button 1`] = `
<div
  class="tui-dropdown tui-dropdown--flexibleWidth"
>
  <div
    class="tui-dropdown__trigger"
    id="id-trigger"
  >
    <div
      class="tui-tagList"
    >
      <div
        class="tui-tagList__tags"
      >
        <div
          aria-atomic="false"
          aria-label="[[selected_tag_list_name, totara_core, \\"Learning category\\"]]"
          aria-live="polite"
          aria-relevant="additions"
          class="tui-tagList__tagItems"
          role="status"
        >
          <div
            class="tui-tag"
          >
            <span
              class="tui-tag__content tui-tag__content--withButton"
            >
              Tui
            </span>
             
            <button
              aria-label="[[tag_remove_from, totara_core, {\\"name\\":\\"Tui\\",\\"taglist\\":\\"Learning category\\"}]]"
              class="tui-iconBtn tui-iconBtn--prim tui-iconBtn--small tui-iconBtn--transparent"
              title="[[tag_remove_from, totara_core, {\\"name\\":\\"Tui\\",\\"taglist\\":\\"Learning category\\"}]]"
              type="button"
            >
              <span
                class="tui-iconBtn__wrap"
              >
                <span
                  class="tui-iconBtn__label"
                >
                  <span
                    aria-hidden="true"
                    class="tui-iconBtn__icon"
                  >
                    <span
                      class=""
                    />
                  </span>
                   
                  <!---->
                </span>
                 
                <!---->
                 
                <!---->
              </span>
            </button>
          </div>
           
          <div
            class="tui-tagList__input"
          >
            <input
              aria-label="[[filter_x_taglist, totara_core, \\"Learning category\\"]]"
              class="tui-formInput tui-formInput--transparent"
              type="text"
            />
          </div>
        </div>
         
        <span
          class="tui-tagList__suffix"
          style="display: none;"
        >
          [[n_more, totara_core, null]]
        </span>
      </div>
       
      <button
        aria-expanded="false"
        aria-haspopup="menu"
        aria-label="[[tag_list_x, totara_core, \\"Learning category\\"]]"
        class="tui-iconBtn tui-tagList__expandArrow tui-iconBtn--transparent"
        title="[[tag_list_x, totara_core, \\"Learning category\\"]]"
        type="button"
      >
        <span
          class="tui-iconBtn__wrap"
        >
          <span
            class="tui-iconBtn__label"
          >
            <span
              aria-hidden="true"
              class="tui-iconBtn__icon"
            >
              <span
                class=""
              />
            </span>
             
            <!---->
          </span>
           
          <!---->
           
          <!---->
        </span>
      </button>
    </div>
  </div>
   
  <!---->
</div>
`;
